delivery.go

```diff
@@ -1,70 +1,122 @@
+	rejectedKey string,
+	pushKey string,
+	redisClient RedisClient,
+	errChan chan<- error,
+) *redisDelivery {
+	return &redisDelivery{
+		ctx:         ctx,
 		payload:     payload,
 		unackedKey:  unackedKey,
 		rejectedKey: rejectedKey,
 		pushKey:     pushKey,
 		redisClient: redisClient,
+		errChan:     errChan,
 	}
 }
 
-func (delivery *wrapDelivery) String() string {
+func (delivery *redisDelivery) String() string {
 	return fmt.Sprintf("[%s %s]", delivery.payload, delivery.unackedKey)
 }
 
-func (delivery *wrapDelivery) Payload() string {
+func (delivery *redisDelivery) Payload() string {
 	return delivery.payload
 }
 
-func (delivery *wrapDelivery) Ack() bool {
-	// debug(fmt.Sprintf("delivery ack %s", delivery)) // COMMENTOUT
+// blocking versions of the functions below with the following behavior:
+// 1. return immediately if the operation succeeded or failed with ErrorNotFound
+// 2. in case of other redis errors, send them to the errors chan and retry after a sleep
+// 3. if redis errors occur after StopConsuming() has been called, ErrorConsumingStopped will be returned
+
+func (delivery *redisDelivery) Ack() error {
+	errorCount := 0
+	for {
+		count, err := delivery.redisClient.LRem(delivery.unackedKey, 1, delivery.payload)
+		if err == nil { // no redis error
+			if count == 0 {
+				return ErrorNotFound
+			}
+			return nil
+		}
+
+		// redis error
+
+		errorCount++
+
+		select { // try to add error to channel, but don't block
+		case delivery.errChan <- &DeliveryError{Delivery: delivery, RedisErr: err, Count: errorCount}:
+		default:
+		}
+
+		if err := delivery.ctx.Err(); err != nil {
+			return ErrorConsumingStopped
+		}
 
-	count, ok := delivery.redisClient.LRem(delivery.unackedKey, 1, delivery.payload)
-	return ok && count == 1
+		time.Sleep(time.Second)
+	}
 }
 
-func (delivery *wrapDelivery) Reject() bool {
+func (delivery *redisDelivery) Reject() error {
 	return delivery.move(delivery.rejectedKey)
 }
 
-func (delivery *wrapDelivery) Push() bool {
-	if delivery.pushKey != "" {
-		return delivery.move(delivery.pushKey)
-	} else {
-		return delivery.move(delivery.rejectedKey)
+func (delivery *redisDelivery) Push() error {
+	if delivery.pushKey == "" {
+		return delivery.Reject() // fall back to rejecting
 	}
+
+	return delivery.move(delivery.pushKey)
 }
 
-func (delivery *wrapDelivery) move(key string) bool {
-	if ok := delivery.redisClient.LPush(key, delivery.payload); !ok {
-		return false
-	}
+func (delivery *redisDelivery) move(key string) error {
+	errorCount := 0
+	for {
+		_, err := delivery.redisClient.LPush(key, delivery.payload)
+		if err == nil { // success
+			break
+		}
+		// error
 
-	if _, ok := delivery.redisClient.LRem(delivery.unackedKey, 1, delivery.payload); !ok {
-		return false
+		errorCount++
+
+		select { // try to add error to channel, but don't block
+		case delivery.errChan <- &DeliveryError{Delivery: delivery, RedisErr: err, Count: errorCount}:
+		default:
+		}
+
+		time.Sleep(time.Second)
```

Comment:
there seems no way to exit this loop in case of error?

Reply:
the `move` method should probably also check for 

```
                if err := delivery.ctx.Err(); err != nil {
			return ErrorConsumingStopped
		}
```

Reply:
Good point, added 👍